security/pkg/nodeagent/sds/sds_stream_test.go

```diff
@@ -0,0 +1,373 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package sds
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	sds "github.com/envoyproxy/go-control-plane/envoy/service/secret/v3"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"k8s.io/apimachinery/pkg/util/uuid"
+
+	"istio.io/istio/security/pkg/nodeagent/cache"
+	"istio.io/istio/security/pkg/nodeagent/model"
+	"istio.io/istio/security/pkg/nodeagent/util"
+)
+
+const (
+	ValidProxyID        = "sidecar~127.0.0.1~SecretsPushStreamOne~local"
+	InValidProxyID      = "invalid~sidecar~127.0.0.1~SecretsPushStreamOne~local"
+	BlockGenSecretError = "BLOCK_GENERATE_SECRET_FOR_TEST_ERROR"
+)
+
+func TestSDSAgentStreamWithCacheAndConnectionCleaned(t *testing.T) {
+	setup := StartStreamTest(t)
+	defer setup.server.Stop()
+
+	conn, stream := createSDSStream(t, setup.socket, fakeToken1)
+	notifyChan := make(chan notifyMsg)
+
+	go testSDSSuccessIngressStreamCache(stream, ValidProxyID, notifyChan)
+	// verify that the first SDS request sent by two streams do not hit cache.
+	waitForStreamSecretCacheCheck(t, setup.secretStore, false, 1)
+	waitForStreamNotificationToProceed(t, notifyChan, "notify push secret 1")
+	conn.Close()
+	// verify the cache is cleaned when connection is closed
+	waitForSecretCacheCleanUp(t, setup.secretStore)
+
+	conn, stream = createSDSStream(t, setup.socket, fakeToken1)
+	go testSDSInvalidIngressStreamCache(stream, InValidProxyID, notifyChan)
+	waitForStreamSecretCacheCheck(t, setup.secretStore, false, 1)
+	waitForStreamNotificationToProceed(t, notifyChan, "notify push secret 2")
+	conn.Close()
+	// verify the cache is cleaned when connection is closed
+	waitForSecretCacheCleanUp(t, setup.secretStore)
+}
+
+func waitForSecretCacheCleanUp(t *testing.T, secretsStore *mockIngressGatewaySecretStore) {
+	waitTimeout := 8 * time.Second
+	start := time.Now()
+	for {
+		cacheCleaned := false
+		secretsStore.secrets.Range(func(key, value interface{}) bool {
+			cacheCleaned = true
```

Comment:
Not clear how this validates cleanup. Can we check if secretsStore.secrets still have an entry with the connID of closed connection?

Reply:
In my purpose, when the connection is created, a secret will be installed into the secret, after the connection is closed. The secret should be deleted from the secret cache and thus  the secrets should contain no secrets anymore. But I will try to extract the cache id and verify it is not in the secrets map

Reply:
Perhaps the name cacheCleaned is a little confusing. When cacheCleaned is false at line 76, !cacheCleaned is true and return. Actually we return from there because cache is cleaned. Perhaps rename it to hasStaledCache, hasStaledCache := false initially, set set it to true inside Range(). if !hasStaledCache return.